patroni/quorum.py

```diff
@@ -0,0 +1,429 @@
+
+        .. note::
+            Merge two transitions of the same type to a single one.
+
+            This is always safe because skipping the first transition is equivalent
+            to no one observing the intermediate state.
+
+        :yields: transitions as :class:`Transition` objects.
+        """
+        transitions = list(self._generate_transitions())
+        for cur_transition, next_transition in zip(transitions, transitions[1:] + [None]):
+            if isinstance(next_transition, Transition) \
+                    and cur_transition.transition_type == next_transition.transition_type:
+                continue
+            yield cur_transition
+            if cur_transition.transition_type == 'restart':
+                break
+
+    def __handle_non_steady_cases(self) -> Iterator[Transition]:
+        """Handle cases when set of transitions produces on previous run was interrupted.
+
+        :yields: transitions as :class:`Transition` objects.
+        """
+        if self.sync < self.voters:
+            logger.debug("Case 1: synchronous_standby_names subset of DCS state")
+            # Case 1: voters is superset of sync nodes. In the middle of changing voters (quorum).
+            # Evict  dead nodes from voters that are not being synced.
+            remove_from_voters = self.voters - (self.sync | self.active)
+            if remove_from_voters:
+                yield from self.quorum_update(
+                    quorum=len(self.voters) - len(remove_from_voters) - self.numsync,
+                    voters=CaseInsensitiveSet(self.voters - remove_from_voters),
+                    adjust_quorum=not (self.sync - self.active))
+            # Start syncing to nodes that are in voters and alive
+            add_to_sync = (self.voters & self.active) - self.sync
+            if add_to_sync:
+                yield from self.sync_update(self.numsync, CaseInsensitiveSet(self.sync | add_to_sync))
+        elif self.sync > self.voters:
+            logger.debug("Case 2: synchronous_standby_names superset of DCS state")
+            # Case 2: sync is superset of voters nodes. In the middle of changing replication factor (sync).
+            # Add to voters nodes that are already synced and active
+            add_to_voters = (self.sync - self.voters) & self.active
+            if add_to_voters:
+                voters = CaseInsensitiveSet(self.voters | add_to_voters)
+                yield from self.quorum_update(len(voters) - self.numsync, voters)
+            # Remove from sync nodes that are dead
+            remove_from_sync = self.sync - self.voters
+            if remove_from_sync:
+                yield from self.sync_update(
+                    numsync=min(self.numsync, len(self.sync) - len(remove_from_sync)),
+                    sync=CaseInsensitiveSet(self.sync - remove_from_sync))
+
+        # After handling these two cases voters and sync must match.
+        assert self.voters == self.sync
```

Comment:
I wonder if throwing a QuorumError instead of AssertionError would look more elegant